test/test.js

```diff
@@ -0,0 +1,406 @@
+
+  const err =
+`TypeError: Invalid value
+
+Point.Point :: Number -> Number -> { x :: Number, y :: Number }
+                         ^^^^^^
+                           1
+
+1)  "foo" :: String
+
+The value at position 1 is not a member of ‘Number’.
+`;
+
+  const Point = Named(
+    'Point'
+    , {Point: {x: Number, y: Number}}
+  );
+
+  try {
+    Point.Point(4, 'foo');
+  } catch (e) {
+    t.equal(err, e.toString());
+  }
+
+  t.end();
+});
+
+test('Switching on union types', t => {
+
+  const Action =
+    Type({
+      Up: [],
+      Right: [],
+      Down: [],
+      Left: [],
+    });
+
+  const player = {x: 0, y: 0};
+
+  const advancePlayer = (action, player) =>
+    Action.case({
+      Up: () => ({x: player.x, y: player.y - 1}),
+      Right: () => ({x: player.x + 1, y: player.y}),
+      Down: () => ({x: player.x, y: player.y + 1}),
+      Left: () => ({x: player.x - 1, y: player.y}),
+    }, action);
+
+  t.deepEqual(
+    {x: 0, y: -1},
+    advancePlayer(Action.Up(), player)
+  );
+
+  t.end();
+
+});
+
+test('Switch on union types point free', t => {
+
+  const Point = Type(
+    {Point: {x: T.Number, y: T.Number}}
+  );
+
+  const Shape = Type({
+    Circle: [T.Number, Point],
+    Rectangle: [Point, Point],
+  });
+
+  const p1 = Point.PointOf({x: 0, y: 0});
+  const p2 = Point.PointOf({x: 10, y: 10});
+
+  {
+    const area = Shape.case({
+      Circle: (radius, _) => Math.PI * radius * radius,
+      Rectangle: (p1, p2) => (p2.x - p1.x) * (p2.y - p1.y),
+    });
+
+    const rect = Shape.Rectangle(p1, p2);
+
+    t.equal(
+      100
+      , area(rect)
+    );
+  }
+
+  {
+    const rect = Shape.Rectangle(p1, p2);
+
+    const area = rect.case({
+      Circle: (radius, _) => Math.PI * radius * radius,
+      Rectangle: (p1, p2) => (p2.x - p1.x) * (p2.y - p1.y),
+    });
+    console.log('after', area);
+
+    t.equal(
+      100
+      , area
+    );
+
+  }
+  t.end();
+});
+
+test('Pass extra args to case via caseOn', t => {
+  const Action =
+    Type({
+      Up: [],
+      Right: [],
+      Down: [],
+      Left: [],
+    });
+
+  const player = {x: 0, y: 0};
+
+  const advancePlayer =
+    Action.caseOn({
+      Up: (p, ...extra) => ['Up', p, ...extra],
+      Right: (p, ...extra) => ['Down', p, ...extra],
+      Down: (p, ...extra) => ['Left', p, ...extra],
+      Left: (p, ...extra) => ['Right', p, ...extra],
+    });
+
+  t.deepEqual(
+    ['Up', {x: 0, y: 0}, 1, 2, 3]
+    , advancePlayer(Action.Up(), player, 1, 2, 3)
+  );
+
+  t.end();
+});
+
+test('Destructuring assignment to extract values', t => {
+
+  const Point = Type(
+    {Point: {x: Number, y: Number}}
+  );
+
+  const [x, y] = Point.PointOf({x: 0, y: 0});
+
+  t.deepEqual(
+    {x: 0, y: 0},
+    {x, y}
+  );
+
+  t.end();
+
+});
+
+test('Recursive Union Types', t => {
+  /* eslint-disable no-var */
+  var List =
+    Type({Nil: [], Cons: [T.Any, List]});
+  /* eslint-enable no-var */
```

Comment:
This is confusing. This would be much clearer:

```javascript
const List = Type({Nil: [], Cons: [T.Any, undefined]});
```

I think we can do better though. How about supporting one of the following?

```javascript
const List = Type({Nil: [], Cons: [T.Any, UnionType.Self]});
```

```javascript
const List = Type({Nil: [], Cons: [T.Any, UnionType.Recur]});
```

```javascript
const List = Type({Nil: [], Cons: [T.Any, UnionType.Recursive]});
```

The value of `UnionType.Self` (or whatever we were to call it) *could* simply be `undefined`, but I'd prefer something like `{'@@functional/recursive-type-reference': true}`.